### ISSUE TITLE
Training llama, loss does not decrease (model does not learn at all)

### ISSUE THREAD

**Issue by KeshavSingh29** [huggingface/optimum-neuron#1019]
### System Info

```shell
Using huggingface neuron AMI on aws trainium 32xl
- `optimum` version: 1.24.0
- `transformers` version: 4.55.4
- Platform: Linux-6.8.0-1036-aws-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.35.3
- PyTorch version (GPU?): 2.8.0+cu128 (cuda availabe: False)
- Tensorflow version (GPU?): not installed (cuda availabe: NA)


Other: 
- neuronx-cc==2.21.18209.0+043b1bf7
- neuronx-distributed==0.15.22404+1f27bddf
- optimum-neuron==0.4.1
- torch-neuronx==2.8.0.2.10.13553+1e4dd6ca
```

### Who can help?

@michaelbenayoun 
I found you had a similar issue and did a bugfix, perhaps you can help please. 
https://github.com/huggingface/optimum-neuron/pull/774 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

Using llama model for training from scratch on my own dataset on aws trainium 32xl .
Model compilation works without issue. 
Minimum code : 

```python
import tiktoken
from transformers import LlamaConfig
from optimum.neuron.models.training.llama.modeling_llama import LlamaForCausalLM
from optimum.neuron.models.training.config import TrainingNeuronConfig

tokenizer = tiktoken.get_encoding("o200k_base")

# Initialize random wt. model with tp=4
trn_config = TrainingNeuronConfig(
        tensor_parallel_size=4,
    )

# Llama3 config -> for a 0.5B param model
cfg = LlamaConfig(
        hidden_size = 1024,
        intermediate_size = 2048,
        num_hidden_layers = 12,
        num_attention_heads = 8,
        num_key_value_heads = 4,
        max_position_embeddings = 1024,
        weight_decay = 0.1,
        vocab_size=tokenizer.n_words,
        pad_token_id=tokenizer.pad_id,
        bos_token_id=tokenizer.bos_id,
        eos_token_id=list(tokenizer.stop_tokens),
        rms_norm_eps=1e-5,
        tie_word_embeddings=False,
        rope_scaling=None,
        rope_theta=10_000.0,
        torch_dtype='bfloat16',
        use_cache=False,
        # attn_implementation='flash_attention_2',
    )

model = LlamaForCausalLM(config=cfg, trn_config=trn_config)

# Custom dataloader and dataset -> I just create long seq. of data so that no padding is needed. 

class FastDataset(IterableDataset):
    """
    Custom Dataset class 
    Input:
        data_files: list of all the files used for training
            Example:
                {
                "train": ["file1","file2" ...],
                "validation": ["file1","file2" ...],
                }
        split: dataset split (train or val)
        chunk_len: Max len of sequence a model can handle
        stride: analogus to window size
        tokenizer: a tiktoken based tokenizer
    """

    def __init__(
        self,
        data_files: dict,
        split: str,
        chunk_len: int,
        stride: int,
        tokenizer: FastTokenizer,
    ):
        self.data: Any = load_dataset(
            "json",
            data_files=data_files,
            streaming=True,
            split=split,
        )
        self.chunk_len = chunk_len
        self.stride = stride
        self.tokenizer = tokenizer

    def __iter__(self):
        # add sequences to buffer -> less padding tokens
        buffer = []
        last_file = None
        for example in self.data:
            current_file = example.get("file_name")

            if current_file is not None and current_file != last_file:
                # logger.info(f"Processing file: {current_file}") -> Shuffled Dataset so no need
                last_file = current_file

            sequence_ids = example["token_ids"]

            # Inject BOS and EOS
            buffer.append(self.tokenizer.bos_id)
            buffer.extend(sequence_ids)
            buffer.append(self.tokenizer.eos_id)

            while len(buffer) >= self.chunk_len:
                chunk = buffer[: self.chunk_len]
                buffer = buffer[self.stride :]  # slide the window
                yield {"input_ids": torch.tensor(chunk, dtype=torch.long)}


class PretrainCollator:
    """
    Collator for variable-length pretraining sequences.
    Pads to the batch’s max length, builds attention masks,
    and uses `ignore_index` for label padding.
    """

    def __init__(self, tokenizer: FastTokenizer, ignore_index: int = -100):
        self.tokenizer = tokenizer
        self.ignore_index = ignore_index
        self.total_seen_samples = 0
        self.total_tokens_seen = 0

    def __call__(self, batch: list[dict[str, Tensor]]) -> dict[str, Tensor]:
        self.total_seen_samples += len(batch)
        self.total_tokens_seen += sum(
            len(item["input_ids"]) for item in batch
        )  # type: ignore

        # 1) collect all input-id sequences
        sequences: list[Tensor] = [item["input_ids"] for item in batch]

        # 2) pad inputs (pad with pad_id) and labels (pad with ignore_index)
        inputs_padded = pad_sequence(
            sequences,
            batch_first=True,
            padding_value=self.tokenizer.pad_id,
        )
        labels_padded = pad_sequence(
            sequences,
            batch_first=True,
            padding_value=self.ignore_index,
        )

        # 3) build attention mask (1 for real tokens, 0 for padding)
        attention_mask = (inputs_padded != self.tokenizer.pad_id).long()

        return {
            "input_ids": inputs_padded,
            "attention_mask": attention_mask,
            "labels": labels_padded,
        }

# pass to trainer and then trainer.train()
```

### Expected behavior

The loss should decrease but it keeps hovering around 12.3 - 12.5. 
I have tried different datasets as well as changing the learning rate / tp_size. 
The issue might be with the loss function
